File: core/src/mindustry/client/navigation/ClientThread.java
package mindustry.client.navigation;

import arc.*;
import arc.math.*;
import arc.util.*;
import arc.util.async.*;
import mindustry.game.*;

import static mindustry.Vars.*;

public class ClientThread implements Runnable {
    private Thread thread = null;
    public TaskQueue taskQueue = new TaskQueue();
    private static final int updateFPS = 60;
    private static final int updateInterval = 1000 / updateFPS;

    public ClientThread() {
        Events.on(EventType.WorldLoadEvent.class, event -> start());
        Events.on(EventType.ResetEvent.class, event -> stop());
        start();
    }

    /** Starts or restarts the thread. */
    private void start() {
        stop();
        taskQueue.clear();
        thread = Threads.daemon(this);
    }

    /** Stops the thread. */
    private void stop() {
        if(thread != null) {
            thread.interrupt();
            thread = null;
        }
    }

    @Override
    public void run() {
        long[] start = new long[]{Time.millis()};
        Timer.schedule(() -> {
            if (Time.timeSinceMillis(start[0]) > 10000) start(); // Restarts on hang
        }, 5, 5);
        while (true) {
            start[0] = Time.millis();
            try {
                if(state != null && state.isPlaying()) taskQueue.run();
            } catch (Exception e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep((long)Mathf.maxZero(updateInterval - Time.millis() + start[0])); // Only run every updateInterval millis
            } catch (InterruptedException e) {
                stop();
                return;
            }
        }
    }
}

File: core/src/mindustry/client/navigation/ClientThread.kt
package mindustry.client.navigation

import arc.*
import arc.util.*
import mindustry.game.EventType.*
import java.util.concurrent.*
import java.util.function.*

object clientThread {
    private var thread: ThreadPoolExecutor? = null

    @JvmStatic
    fun queue() = thread?.queue

    init {
        Events.on(WorldLoadEvent::class.java) { start() }
//        Events.on(ResetEvent::class.java) { stop() } FINISHME: Breaks navigation obstacles
        start()
    }

    /** Starts or restarts the thread. */
    private fun start() {
        stop()
        thread = ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS, LinkedBlockingQueue()) { r ->
            Thread(r, "Client Thread").apply { isDaemon = true }
        }
    }

    /** Stops the thread. */
    private fun stop() {
        val start = Time.millis()
        Threads.await(thread ?: return) // FINISHME: We can't just shutdownNow as we need to ensure that turret ents are properly added and removed but we also don't want to wait on this for ages for things like chat schem sharing
        Log.debug("Waited @ms for client thread", Time.timeSinceMillis(start))
//        thread?.shutdownNow()
        thread = null
    }

    @JvmStatic
    fun post(task: Runnable): CompletableFuture<Void> = if (thread != null) CompletableFuture.runAsync(task, thread).exceptionHandler() else {
        Log.warn("Post to null clientThread")
        CompletableFuture()
    }

    @JvmStatic
    fun <T> submit(task: Supplier<T>): CompletableFuture<T> = if (thread != null) CompletableFuture.supplyAsync(task, thread).exceptionHandler() else {
        Log.warn("Submit to null clientThread")
        CompletableFuture()
    }

    @JvmStatic
    fun get() = thread

    private fun <T> CompletableFuture<T>.exceptionHandler() = apply { exceptionally {
        Core.app.post {
            if (it is CancellationException) return@post
            Log.err("Exception on client thread", it)
        }
        throw it
    } }
}